core/server/master/src/main/java/alluxio/master/meta/PathProperties.java

```diff
@@ -12,41 +12,235 @@
 package alluxio.master.meta;
 
 import alluxio.conf.PropertyKey;
+import alluxio.master.journal.CheckpointName;
+import alluxio.master.journal.DelegatingJournaled;
+import alluxio.master.journal.JournalContext;
+import alluxio.master.journal.Journaled;
+import alluxio.proto.journal.Journal;
+import alluxio.proto.journal.Journal.JournalEntry;
+import alluxio.proto.journal.Meta.PathPropertiesEntry;
+import alluxio.proto.journal.Meta.RemovePathPropertiesEntry;
+import alluxio.resource.LockResource;
 
 import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
 import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Supplier;
 
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
 import javax.annotation.concurrent.ThreadSafe;
 
 /**
  * Source of truth for path level properties.
- * TODO(cc): journaling
- * TODO(cc): more operations like batch operation, delete, etc
+ *
+ * We assume that operations for path level properties are not highly concurrent,
+ * based on this assumption,
+ * thread safety in this class is guaranteed by a read write lock,
+ * also for simplicity, for an operation on a path, we journal the full properties of the path
+ * after the operation is done.
+ *
+ * If the assumption is incorrect, we need to use a more fine-grained concurrency control method
+ * and only journal the changes to the path properties.
  */
 @ThreadSafe
-public final class PathProperties {
+public final class PathProperties implements DelegatingJournaled {
+  /** Lock for mState. */
+  private final ReadWriteLock mLock = new ReentrantReadWriteLock();
+  /** Journaled state of path level properties. */
+  @GuardedBy("mLock")
+  private final State mState = new State();
+
   /**
-   * Map from path to key value properties.
+   * @return a copy of path level properties which is a map from path to property key values
    */
-  private final ConcurrentHashMap<String, ConcurrentHashMap<PropertyKey, String>> mProperties =
-      new ConcurrentHashMap<>();
+  public Map<String, Map<PropertyKey, String>> get() {
+    try (LockResource r = new LockResource(mLock.readLock())) {
+      Map<String, Map<PropertyKey, String>> copy = new HashMap<>();
+      mState.getProperties().forEach((path, kv) -> {
```

Comment:
can we do the copy inside `mState.getProperties()`, and just return that from here? That way, the mutable map can be isolated within `State`, and the rest of `PathProperties` can't accidentally modify it.

Reply:
I moved the implementation of these methods to `State`, then these methods become delegate to those in `State`